llama.cpp

```diff
@@ -11322,30 +11372,13 @@ static int llama_decode_internal(
             // no output
             res  = nullptr;
             embd = nullptr;
-        } else if (!hparams.causal_attn) {
-            res = nullptr; // do not extract logits for embedding models such as BERT
-
-            // token or sequence embeddings
-            embd = gf->nodes[gf->n_nodes - 1];
-
-            GGML_ASSERT(strcmp(embd->name, "result_embd") == 0 || strcmp(embd->name, "result_embd_pooled") == 0);
         } else if (cparams.embeddings) {
-            // the embeddings could be in the second to last tensor, or any of the previous tensors
-            int i_embd = gf->n_nodes - 2;
-            for (int i = 3; strcmp(embd->name, "result_norm") != 0; ++i) {
-                i_embd = gf->n_nodes - i;
-                if (i_embd < 0) { break; }
-                embd = gf->nodes[i_embd];
-            }
-            GGML_ASSERT(i_embd >= 0 && "missing result_norm tensor");
-
-            // TODO: use a per-batch flag to know when to skip logits while keeping embeddings
-            if (!cparams.causal_attn) {
-                res = nullptr; // do not extract logits when not needed
-                // skip computing logits
-                // TODO: is this safe?
-                gf->n_nodes = i_embd + 1;
+            res = nullptr; // do not extract logits for embedding case
+            embd = gf->nodes[gf->n_nodes - 1];
+            if (strcmp(embd->name, "result_embd_pooled") != 0) {
+                embd = gf->nodes[gf->n_nodes - 2];
             }
+            GGML_ASSERT(strcmp(embd->name, "result_embd_pooled") == 0 && "missing embeddings tensor");
         } else {
             embd = nullptr; // do not extract embeddings when not needed
             GGML_ASSERT(strcmp(res->name, "result_output") == 0 && "missing result_output tensor");
```

Comment:
So an embeddings model will crash on the first decode when `cparams.embeddings` is set to `false`?

Reply:
Yeah, though I can't think of any case where you'd use an embedding model without `cparams.embeddings`. I guess there's nothing really indicating something is an embedding model other than the lack of a `result_output` tensor, so it's hard to intercept this earlier and give an error.

Reply:
Well, `hparams.causal_attn` is `false` for BERT at least, and it's the only embedding-only model architecture currently in `llama.cpp`. All BERT-like architectures also set this key to `false` when converted to GGUF.

There might be a need for a dedicated metadata key-value pair for embedding-only models if non-causal text generation models are a thing. (T5? Or is it causal?) Anyway, `cparams.causal_attn` can be used to get non-causal attention with any model, I think (I did not test this), except for recurrent models (Mamba).

I think there should at least be some abstraction (exported in `llama.h`) to know whether or not a model **can** provide embeddings and/or logits. This would make things like #7448 easier, even if it initially relies on `hparams.causal_attn`.

Reply:
I see, so at least for now, it looks like `hparams.causal_attn` is a good indicator of whether a model is embedding-only. And I can't imagine a generative model with non-causal attention. I think T5 is causal, at least for the decoder part.

Then I guess we want to assert `hparams.causal_attn || cparams.embeddings` at some point. That way we don't have to worry about divergence and the error is caught earlier.